dpu-tools

```diff
@@ -0,0 +1,278 @@
+        )
+    elif detected.out == DPUType.IPU.name:
+        firmware_up_parser.add_argument("--version", choices=VERSIONS)
+
+    firmware_subparsers.add_parser("version", help="Get firmware version").set_defaults(
+        subcommand="firmware_version"
+    )
+
+    subparsers.add_parser("list", help="List devices").set_defaults(
+        subcommand="list_dpus"
+    )
+
+    console_parser = subparsers.add_parser("console", help="Open IPU console")
+    console_parser.set_defaults(subcommand="console")
+    console_parser.add_argument("--target", choices=["imc", "acc"], default="imc")
+
+    if detected.out == DPUType.BF.name:
+        mode_parser = subparsers.add_parser(
+            "mode", help="Retrieves the current mode for a Bluefield"
+        )
+        mode_parser.add_argument("--set-mode", choices=["dpu", "nic"])
+        mode_parser.set_defaults(subcommand="mode")
+        mode_parser.add_argument("--next-boot", action="store_true")
+
+        url = "http://download.eng.brq.redhat.com/released/rhel-9/RHEL-9/9.2.0/BaseOS/aarch64/os/images/efiboot.img"
+        pxeboot_parser = subparsers.add_parser("pxeboot", help="Boots BF using pxeboot")
+        pxeboot_parser.set_defaults(subcommand="pxeboot")
+
+        pxeboot_parser.add_argument(
+            "iso", metavar="iso", type=str, help="iso to use for PXE booting"
+        )
+        pxeboot_parser.add_argument(
+            "-e",
+            metavar="efiboot_img",
+            dest="efiboot_img",
+            default=url,
+            type=str,
+            help="Specify url to pull efiboot from.",
+        )
+        pxeboot_parser.add_argument(
+            "-m",
+            "--wait-minicom",
+            dest="wait_minicom",
+            default=False,
+            action="store_true",
+            help="Instead of running minicom to select pxe entry, just wait indefinitely.",
+        )
+        help = (
+            "if key is specified, the script will log in to the BF and"
+            " run 'ip --json a' before quitting. This also means that the"
+            " script will only block when the BF has booted"
+        )
+        pxeboot_parser.add_argument(
+            "-w", metavar="key", dest="key", default="", type=str, help=help
+        )
+
+    utils_parser = subparsers.add_parser(
+        "utils", help="Other non-dpu utilities that may still be useful"
+    )
+    utils_subparser = utils_parser.add_subparsers(
+        title="utils subcommand", help="Houses the useful non-dpu utilities"
+    )
+    cx_fwup_parser = utils_subparser.add_parser(
+        "cx-fwup", help="Used for upgrading Nvidia cx Smart Nics"
+    )
+    cx_fwup_parser.set_defaults(subcommand="cx-fwup")
+    bfb_parser = utils_subparser.add_parser(
+        "bfb", help="Downloads BFB images and sends it to BF"
+    )
+    bfb_parser.set_defaults(subcommand="bfb")
+    # Parse arguments and initialize DPUTools
+    args = parser.parse_args()
```

Comment:
It might be nice to have this main() function call out to another function to set up all the above arguments. The way it is now makes the code a bit harder to read / harder to track where subsequent changes need to be added.

So main would essentially do:
- set up arguments
-  set up logging
-  create dpuTools
- dispatch dpuTools

Any then the more granular argument organization could go in the "set up arguments" function.